[Spring 체스 - 2단계] 페퍼(최수연) 미션 제출합니다.

SuyeonChoi opened the pull request:
안녕하세요 닉! 페퍼입니다. 1단계 리뷰를 반영하고 2단계 기능을 구현하였습니다.

현재 GameService에서 GameDao와 BoardDao 두가지를 모두 관리하는데 너무 많은 책임을 갖고 있다는 생각이 들어요.
하지만 DB에서도 piece테이블이 game 테이블을 참조하기 때문에 하나의 서비스 클래스에서 일단 구현했습니다.
이런 경우 Service의 책임을 어떻게 분리할 수 있나요..? DB 설계가 잘못된 걸까요?🤔🤔

늦었지만 많은 피드백 부탁드리겠습니다! 
미리 감사드립니다 닉! 🙇‍♀️

SuyeonChoi committed:
refactor(web/GameController): 불필요한 코드 삭제

SuyeonChoi committed:
refactor(application/web): Controller 분리
- 데이터를 반환하는 RestController, View를 반환하는 Controller 분리

SuyeonChoi committed:
refactor(application/web): ControllerAdvice 분리

SuyeonChoi committed:
docs(README.md): 기능 목록 업데이트

SuyeonChoi committed:
feat: 체스방 생성
- sql 스크립트 업데이트

SuyeonChoi committed:
feat: 체스방 조회

SuyeonChoi committed:
feat: 체스방 삭제

SuyeonChoi committed:
feat: 체스방 입장 시 게임 이어서 진행

SuyeonChoi committed:
fix: 기존에 동작하지 않는 기능 해결
- 체스 기물 이동 시 오류 해결
- 체스 방이 삭제되지 않는 오류 해결

SuyeonChoi committed:
test: 테스트가 모두 통과되도록 수정

SuyeonChoi committed:
refactor: 사용하지 않는 코드 삭제

SuyeonChoi committed:
refactor(dao): 메서드가 한가지 행위만 하도록 변경
- 네이밍 변경: load -> find

SuyeonChoi committed:
test(dao): 테스트 코드 추가
- GameDaoTest: findById
- BoardDaoTest: findAllByGameId

SuyeonChoi committed:
refactor: POST 요청시 ajax를 사용하도록 변경

SuyeonChoi committed:
refactor: POST 요청을 RestController가 처리하도록 변경

SuyeonChoi committed:
feat: 게임이 진행 중인 경우 삭제할 수 없는 기능 구현

wbluke commented on src/main/java/chess/application/web/GameService.java:
```diff
@@ -147,4 +155,17 @@ public boolean isGameFinished() {
+        boardDao.deleteAllByGameId(id);
+        gameDao.deleteById(id);
+    }
 }
```
PR에 메서드가 안보여서 여기에 남겨요.
- `public void move(Request req)` 사용하는 메서드일까요? 이것 말고도 사용안하는 메서드가 많은 것 같아요.

SuyeonChoi replied:
전반적으로 service 클래스를 리팩토링하면서 필요하지 않은 코드를 제거했습니다.

wbluke commented on docker/db/mysql/init/init.sql:
```diff
@@ -4,7 +4,10 @@ drop table if exists game cascade;
 create table game
 (
     no         int     not null auto_increment,
```
보통 Auto Increment PK는 Long 타입으로 잡습니다. 자동채번되는 값이고 데이터 삭제와 상관 없이 감소하지 않기 때문에 int형은 작을 수 있거든요.
관련해서 코드도 같이 수정해보면 좋을 것 같아요 :)

SuyeonChoi replied:
저장되는 데이터가 43억개를 넘어가지 않겠다는 보장이 없고, 서버 운영 효율이 낮아지더라도 추후 int ➡️ bigint로 형변환을 하는 비용보다 좋기 때문에 DB의 id 값으로도 주로 `BIGINT`를 사용하는군요!
테이블 구조도 수정하고, 이를 코드에서는 `Long`타입으로 받도록 반영하였습니다.

wbluke commented on src/main/java/chess/application/web/ChessControllerAdvice.java:
```diff
@@ -0,0 +1,14 @@
+
+@ControllerAdvice
+public class ChessControllerAdvice {
+    @ExceptionHandler
```
모든 Exception을 하나의 Handler로만 처리하려는 의도이실까요? :)

SuyeonChoi replied:
비즈니스 로직에서 던지는 `IllegalArgumentException`을 잡기 위하여 선언하였는데 목적대로 구현하지 못했네요😅
인자로 IllegalArgumentException을 전달하여 의도한 예외를 처리할 수 있도록 변경하였습니다. 추가적으로 의도하지 않은 예외가 발생했을 때, 클라이언트에서 날것의 예외를 받지 않도록 `RuntimeException`도 처리하였습니다!

wbluke commented on src/main/java/chess/application/web/GameController.java:
```diff
@@ -1,81 +1,53 @@
 
-    public static void main(String[] args) {
-        SpringApplication.run(GameController.class, args);
+    @PostMapping("/create")
```
[REST API](https://meetup.toast.com/posts/92)에 대한 학습 후 URL 설계를 다시 시도해보면 좋을 것 같아요 :)
(모든 API 대상)

SuyeonChoi replied:
감사합니다! 닉이 링크로 준 자료를 참고해서 공부하고, 이를 `GameController`에 반영하였습니다. 기본적인 Rest Api 규칙을 지키지 않고 있었네요😅
리팩토링을 하며 응답 객체와 요청 객체를 적절한 dto를 생성하여 매핑하였습니다. 그런데 `ViewController`는 뷰 페이지를 함께 반환하다보니 URI를 적절한 값으로 생성한 것인지, 모델에 담아서 페이지를 반환하는 것이 과연 좋은 방법인지 확신이 들지 않습니다... 혹시 ViewController를 더 개선할 방법이 있을까요?

wbluke commented on src/main/java/chess/application/web/GameService.java:
```diff
@@ -69,14 +66,23 @@ public Map<String, Object> modelPlayingBoard() {
+        return gameId;
+    }
+
+    public Map<String, Object> findBoardByGameId(int id) {
```
반환형이 꼭 Map이어야 할까요?
RestController를 사용해 JSON으로 소통하는 만큼 모든 API에 대해 DTO를 사용한 요청/응답으로 설계해보면 좋을 것 같아요.
DTO가 생기면 그에 따른 최소한의 검증이나 가공도 가능해지거든요 :)

SuyeonChoi replied:
이전에는 뷰를 반환할 때, model에 데이터를 넣기 위해서는 Map 객체만 가능하다고 잘못 알고 있었습니다.
model을 함께 반환해도 충분히 객체로 반환할 수 있다는 것을 확인하고, API가 적절한 응답 dto 객체를 요청받고 응답하도록 구현하였습니다!

wbluke commented on src/main/java/chess/application/web/GameService.java:
```diff
@@ -47,14 +50,8 @@ public GameService(GameDao gameDao, BoardDao boardDao) {
         this.boardDao = boardDao;
```
> 현재 GameService에서 GameDao와 BoardDao 두가지를 모두 관리하는데 너무 많은 책임을 갖고 있다는 생각이 들어요.
하지만 DB에서도 piece테이블이 game 테이블을 참조하기 때문에 하나의 서비스 클래스에서 일단 구현했습니다.
이런 경우 Service의 책임을 어떻게 분리할 수 있나요..? DB 설계가 잘못된 걸까요?🤔🤔

GameService에서 GameDao와 BoardDao를 관리한다는 관점이 아니라 필요한 Data Access Object(DAO)를 가져가다 사용한다는 관점으로 보면 좋을 것 같아요.
하나의 Service에서 사용하거나 접근하는 엔티티가 많을수록 충분히 여러 DAO를 참조할 수 있거든요.
(물론 Service의 역할만 보았을 때 2가지 테마의 일을 하고 있다면 당연히 Service를 분리해야겠죠?)
반대로 하나의 DAO가 필요에 따라 여러 Service에서 사용될 수도 있구요!

그래서 현재 Service가 복잡해보이는건 DAO를 2가지나 관리/사용하고 있어서가 아니라, 각 요청에서 `도메인 로직의 수행`과 `데이터의 가공`이 혼재되어 있기 때문인 것 같아요 :)

SuyeonChoi replied:
닉 자세한 리뷰 감사합니다!
리뷰를 보고 `GameService`를 살펴보니, 아무래도 인스턴스 필드인 `chessGame`을 상태처럼 이용하기 때문에 서비스가 혼란스러워진다고 판단하였습니다. 그래서 해당 필드를 제거하고, 서비스에서 `dao`만 관리하도록 변경하였더니 보다 깔끔해진거 같아요!
추가적으로 기존에 `dao`에서는 `dto`를 데이터로 받았는데, `dto`가 변경이 될때마다 `dao` 코드까지 변경해야해서 혼란스럽더라구요. 이 부분은 테이블 구조와 똑같은 entity 클래스를 만들어서 dao에서 보다 편리하게 데이터를 관리하도록 바꾸어보았습니다!

wbluke commented on src/test/java/chess/dao/GameDaoTest.java:
```diff
@@ -1,33 +1,87 @@
+        chessGame = new ChessGame("test", "test");
+        id = gameDao.save(chessGame);
+    }
+
```
보통 이런 Test Fixture를 준비할 때에는 (중복이더라도) 각 테스트에서 만들어주는 것이 좋습니다.
테스트 하나를 이해하기 위해 항상 상단에 있는 테스트 조건까지 같이 확인해야 하는데, 
테스트의 중요한 기능 중 하나가 바로 비즈니스 행위를 이해할 수 있는 `문서`기 때문에 이와 같은 작성법은 해당 기능을 비효율적으로 만들게 됩니다.
반대로 직접적인 테스트 대상은 아니나 테스트를 위해 반복적으로 생성해야 하는 부가적인 데이터라면 setUp()에 위치시킬 수도 있겠습니다만, 저는 개인적으로 좋아하지는 않네요 ㅎㅎ (아무리 부가적인 데이터여도 위에 말한 이유 때문에요)

그리고 `given-when-then`에 대해서 학습하고 적용해보신다면 조금 더 감이 오실 수 있을 것 같아요 :)

SuyeonChoi replied:
`@BeforeEach`와 `@AfterEach`는 테스트 전/후 같은 환경을 구성하도록 사용해야하는데, 반복되는 코드를 넣다 보니 의도와 맞지 않게 구현했었네요😅😅
닉의 조언대로 `given-when-then`에 대해 학습하고 테스트에 이를 적용해보았습니다. 비록 코드는 중복되더라도, 테스트간 독립성을 보장할 수 있어 보다 나은 테스트 코드로 만들었습니다👍

wbluke commented on src/test/java/chess/dao/GameDaoTest.java:
```diff
@@ -1,33 +1,87 @@
+        List<GameDto> result = gameDao.findAll();
+
+        assertThat(result).isNotEmpty();
+    }
```
보통 이런 CRD 테스트가 애매한데요.
저장을 하면 조회를 해서 확인해야 하고, 그 이후에 다른 테스트를 위해 데이터 클렌징도 겸하기 때문입니다.
저는 그래서 보통 CR이나 CRD 테스트를 하나의 테스트로 묶어서 진행합니다.
저장하고, 조회해서 정보를 확인하고, 그 이후에 삭제한 뒤 다시 재조회하여 삭제가 잘 되었는지까지 한번에 검증하곤 해요.

SuyeonChoi replied:
닉의 조언대로, CR로 묶어 테스트를 진행하고, `@AfterEach`를 통해 테스트 DB의 데이터를 모두 제거하는 방식으로 진행해보았습니다. 이전에는 억지로 통과되는 테스트 코드를 작성하는 느낌이었는데, 통일된 시나리오가 있으니 데이터의 삽입, 조회, 삭제를 정확하게 검증할 수 있다고 생각했습니다!

wbluke commented on src/test/java/chess/dao/GameDaoTest.java:
```diff
@@ -1,33 +1,87 @@
+    void findAll() {
+        List<GameDto> result = gameDao.findAll();
+
+        assertThat(result).isNotEmpty();
```
모든 데이터를 조회했다면 어떤 데이터가 있어야 정상인지도 다 확인해야 올바른 테스트일 것 같아요.
저는 Junit5에서 다음 구문을 즐겨 사용합니다 :)
```java
assertThat(results).hasSize(1) // 사이즈 체크를 하면서 다음 검증 체이닝
        .extracting("name", "age") // 추출해서 비교하고자 하는 필드명
        .containsExactly(...); // 순서가 보장되지 않는다면 containsExactlyInAnyOrder
```

SuyeonChoi replied:
생각하지 못한 부분이네요.. 단순히 통과되는 테스트를 만들던 지난 날을 반성합니다,,,🥲

wbluke commented on src/test/java/chess/dao/BoardDaoTest.java:
```diff
@@ -1,50 +1,57 @@
     BoardDao boardDao;
     @Autowired
     GameDao gameDao;
+    int gameId;
```
필드에 의존하는 테스트가 생기면 각 테스트가 서로를 참조할 수 있기 때문에 테스트 간 독립성이 보장되지 않을 수 있습니다.
GameDaoTest 쪽에도 코멘트 남겼지만 각 테스트마다 Fixture를 새로 만들어서 독립성이 보장되도록 해주세요 :)

wbluke commented on src/main/java/chess/application/web/GameService.java:
```diff
@@ -147,4 +155,17 @@ public boolean isGameFinished() {
     }
+
+    public void deleteGame(int id, Map<String, String> request) {
+        String password = request.get("password");
```
Request가 DTO라면 이런 매직스트링도 해결이 될 것 같네요 :)
위에 있는 상수들도 많이 정리가 될 것 같고요.

SuyeonChoi replied:
확실히 Map객체보다는 DTO객체를 사용하는 것이 많은 장점이 있네요...!

wbluke commented on src/main/java/chess/application/web/GameService.java:
```diff
@@ -147,4 +155,17 @@ public boolean isGameFinished() {
         return chessGame.getBoardSquares();
     }
+
+    public void deleteGame(int id, Map<String, String> request) {
```
Service 테스트는 없나요?
DAO 테스트만 있고 정작 더 중요한 도메인 기능에 대한 테스트가 누락된 것 같아요! :)

wbluke reviewed (changes requested):
2단계도 잘 구현해 주셨네요 ㅎㅎ
본격적으로 리팩토링 더 진행해보면 좋을 것 같아서, 여러 코멘트 남겼어요.
확인해보시고 재요청 부탁드려요 🙂

SuyeonChoi committed:
refactor: id 타입 int -> Long 변경
- 테이블의 PK 칼럼명 no -> id로 변경

SuyeonChoi committed:
feat(ChessControllerAdvice): IllegalArgumentException 예외를 처리할 수 있도록 변경
- RuntimeException 처리 추가
- 예외 반환 시, String이 아닌 ErrorResponse 객체를 응답하도록 변경

SuyeonChoi committed:
refactor(GameController): rest api 디자인 가이드를 지키도록 수정
- 게임 진행 상태에서 삭제하는 경우, IllegalArgumentException -> IllegalArgumentException 발생으로 변경
- move마다 기물을 저장하므로, 임의로 게임 상태를 저장하는 기능 제거

SuyeonChoi committed:
refactor: api가 객체를 응답하도록 변경
- controller) Model과 View를 모두 반환하는 메서드는 ModelAndView 객체를 리턴하여 model에 json 객체를 담도록 변경
- service) 메서드 재정렬 및 사용하지 않는 코드 제거

SuyeonChoi committed:
refactor(GameService): ChessGame를 필드에서 제거하여 상태를 갖지 않도록 변경
- DB테이블에 매핑되는 entity를 생성하여 dao에서 dto 대신 데이터를 받도록 변경
- test(dao): given-when-then 시나리오에 부합하는 테스트 작성
- test(GameDaoTest): updateStateById 테스트 추가
- test(BoardDaoTest): deleteAllByGameId 테스트 추가
- rename: GameDto -> GameResponse
- refactor(ChessGame): title, password 인스턴스 필드 제거

SuyeonChoi committed:
test: GameService에 대한 테스트코드 작성

SuyeonChoi reviewed (commented):
안녕하세요 닉!
미리 말씀드렸는데 코로나로 거의 일주일을 날리고....😭
늦었지만 많은 공부를 하며 피드백을 반영해보았습니다 :)
공부하며 의문이 드는점 하나가 있어 코멘트로 남겨놓았습니다!

많이 늦었지만 잘 부탁드리겠습니다!

SuyeonChoi commented on src/test/java/chess/dao/BoardDaoTest.java:
```diff
@@ -1,51 +1,91 @@
 
-        boardDao.save(squares);
-        List<PieceDto> board = boardDao.load();
+        jdbcTemplate.update(connection -> connection.prepareStatement(sql, new String[]{"id"}), keyHolder);
 
-        PieceDto pieceAtA1 = board.stream()
-                .filter(pieceDto2 -> pieceDto2.getPosition().equals("a1"))
-                .findAny().get();
-        assertThat(pieceAtA1.getCamp() + pieceAtA1.getType()).isEqualTo("whiterook");
+        return Objects.requireNonNull(keyHolder.getKey()).longValue();
```
해당 코드는 `GameDao`클래스의 `save`메서드와 거의 똑같은 코드입니다. 하지만 `BoardDaoTest`에서 `GameDao`를 갖는게 어색해서 추가적으로 중복되는 코드를 테스트 클래스에 작성해보았습니다. 저는 중복 코드가 생기지만, `BoardDaoTest`에서 `GameDao`를 의존하지 않는 점은 좋다고 판단했어요! 닉은 어떻게 생각하시나요?

wbluke commented on src/test/java/chess/dao/BoardDaoTest.java:
```diff
@@ -2,57 +2,90 @@
+        assertThat(boardDao.findAllByGameId(gameId)).isEmpty();
+    }
+
+    private Long saveGame() {
```
> 해당 코드는 GameDao클래스의 save메서드와 거의 똑같은 코드입니다. 하지만 BoardDaoTest에서 GameDao를 갖는게 어색해서 추가적으로 중복되는 코드를 테스트 클래스에 작성해보았습니다. 저는 중복 코드가 생기지만, BoardDaoTest에서 GameDao를 의존하지 않는 점은 좋다고 판단했어요! 닉은 어떻게 생각하시나요?

음 네네 ㅎㅎ 하지만 반대로 생각해보면 GameDao 관련 로직이 수정된다면 모든 부분이 각각 수정되어야 할텐데요.
저라면 GameDao의 기능을 사용할 것 같아요 :)

wbluke reviewed (approved):
긴 기간동안 고생 많으셨어요 ㅎㅎ
남은 레벨2 미션도 화이팅입니다! 🙂
